src/hotspot/share/opto/node.hpp

```diff
@@ -578,6 +578,12 @@ class Node {
     _in[i2] = n1;
     // If this node is in the hash table, make sure it doesn't need a rehash.
     assert(check_hash == NO_HASH || check_hash == hash(), "edge swap must preserve hash code");
+    // Flip swapped edges flag.
+    if (has_swapped_edges()) {
+      remove_flag(Node::Flag_has_swapped_edges);
+    } else {
+      add_flag(Node::Flag_has_swapped_edges);
+    }
```

Comment:
This is probably the riskiest part of the implementation. If anyone decides to swap edges without using `swap_edges`, we will not detect the reduction. We have to trust our tests to catch that.

Reply:
That's right, in that case we would miss the reduction. On the other hand, the assumption that edges are ordered in the current form is deeply ingrained in C2, and if this was altered, missing some reduction vectorizations would be the least of our problems. And, as you mention, thanks to the introduction of IR checks in the reduction tests we would detect the missing vectorization anyway.